src/frequenz/client/base/client.py

```diff
@@ -5,14 +5,102 @@
 
 import abc
 import inspect
+import secrets
+import time
+from base64 import urlsafe_b64encode
 from collections.abc import Awaitable, Callable
 from typing import Any, Generic, Self, TypeVar, overload
 
-from grpc.aio import AioRpcError, Channel
+from cryptography.hazmat.primitives.hashes import SHA256
+from cryptography.hazmat.primitives.hmac import HMAC
+from grpc.aio import (
+    AioRpcError,
+    Channel,
+    ClientCallDetails,
+    Metadata,
+    UnaryUnaryCall,
+    UnaryUnaryClientInterceptor,
+)
 
 from .channel import ChannelOptions, parse_grpc_uri
 from .exception import ApiClientError, ClientNotConnected
 
+RequestT = TypeVar("RequestT")
+ResponseT = TypeVar("ResponseT")
+"""The request and response types of a gRPC method."""
+
+
+# UnaryUnaryClientInterceptor is weird from a typing perspective. It is impossible
+# to make both mypy and the runtime happy at the same time. ChatGPT will hallucinate.
+# You have been warned.
+class MetadataHMACInterceptor(UnaryUnaryClientInterceptor):  # type: ignore
+    """An Interceptor that adds HMAC authentication of the metadata fields to a gRPC call."""
+
+    def __init__(self, key: str, secret: str):
+        """Create an instance of the interceptor.
+
+        Args:
+            key: The API key used for HMAC authentication.
+            secret: The API secret used for HMAC authentication.
+        """
+        self._key = key.encode()
+        self._secret = secret.encode()
```

Comment:
I would use keyword-only arguments here, being strings it could be easy to mix up, and name the `key` auth_key` for extra clarity (nitpick, reverse the order too, as the secret is really **the** thing for singing, `auth_key` is just to add to the signature, and clarify this in the docs):
```suggestion
    def __init__(self, *, secret: str, auht_key: str, ):
        """Create an instance of the interceptor.

        Args:
            secret: The secret used to sign the messages using HMAC.
            auth_key: The API key used to authenticate to the service, used as metadata to build the signature.
        """
        self._secret = secret.encode()
        self._auth_key = auth_key.encode()
```

Reply:
Gladly.

Reply:
Using keyword only arguments now.